src/com/spike/ErrorReporter.java

```diff
@@ -0,0 +1,57 @@
+package com.spike;
+
+import java.io.OutputStream;
+import java.io.PrintWriter;
+
+// TODO: Make errors great again
+class ErrorReporter {
+  private final PrintWriter writer;
+  private boolean hadErrors;
+  private final String[] lines;
+
+  class Exception extends RuntimeException {
```

Comment:
Consider renaming `Exception` class to `ReportException`. Redefition of root `java.lang.Exception` is not good practice. Also consider to move it to separate file, so when we'll have many different exceptions we could combine them in one module `com.spike.exceptions`.

Reply:
I thought about that too. The motivation is that it will be used as `ErrorReporter.Exception`. Otherwise there would be too many 'reporters' for one name.

Reply:
What do you think about moving it out to separate file?
`ReportException` is shorter then `ErrorReporter.Exception`. Also check this link, maybe of some value https://stackoverflow.com/questions/19326615/java-defining-custom-exceptions-inside-of-class-is-this-bad

Reply:
Btw, what about making this exception checked?

Reply:
Haven't got it, checked exception is an exception that cab be checked at compile time, but ReportException depends on file contents therefore it is unchecked, i.e. RuntimeException, so current implementation is good.

Reply:
The idea is that we will be forced to add `throws` clause and won't miss the point where we should catch an exception (and we should catch it anyway since it is used for synchronization)

Reply:
`ReportException` is raised at runtime if user provides bad input, so it is unchecked by nature (similar to `NumberFormatException`). If you want to annotate with `throws`, as I remember you can do it for any exception.

Reply:
This violates the idea of checked/unchecked exceptions.